addons/l10n_it_edi_doi/models/sale_order.py

```diff
@@ -0,0 +1,238 @@
+    )
+
+    l10n_it_edi_doi_id = fields.Many2one(
+        string="Declaration of Intent",
+        compute='_compute_l10n_it_edi_doi_id',
+        store=True, readonly=False, precompute=True,
+        comodel_name='l10n_it_edi_doi.declaration_of_intent',
+    )
+
+    l10n_it_edi_doi_not_yet_invoiced = fields.Monetary(
+        string='Declaration of Intent Amount Not Yet Invoiced',
+        compute='_compute_l10n_it_edi_doi_not_yet_invoiced',
+        store=True, readonly=True,
+        help="Total under the Declaration of Intent of this document that can still be invoiced",
+    )
+
+    l10n_it_edi_doi_warning = fields.Text(
+        string="Declaration of Intent Threshold Warning",
+        compute='_compute_l10n_it_edi_doi_warning',
+    )
+
+    @api.depends('commitment_date')
+    def _compute_l10n_it_edi_doi_date(self):
+        for order in self:
+            order.l10n_it_edi_doi_date = order.commitment_date or fields.Date.context_today(self)
+
+    @api.depends('l10n_it_edi_doi_id', 'country_code')
+    def _compute_l10n_it_edi_doi_use(self):
+        for order in self:
+            order.l10n_it_edi_doi_use = order.l10n_it_edi_doi_id \
+                or "IT" in order.country_code
+
+    @api.depends('company_id', 'partner_id.commercial_partner_id', 'l10n_it_edi_doi_date', 'currency_id')
+    def _compute_l10n_it_edi_doi_id(self):
+        for order in self:
+            if order.state != 'draft' or not order.l10n_it_edi_doi_use:
+                order.l10n_it_edi_doi_id = False
+                continue
+
+            company = order.company_id
+            partner = order.partner_id.commercial_partner_id
+            date = order.l10n_it_edi_doi_date
+            currency = order.currency_id
+
+            # Avoid a query or changing a manually set declaration of intent
+            # (if the declaration is still valid).
+            current_declaration = order.l10n_it_edi_doi_id
+            if current_declaration and not current_declaration._get_validity_warnings(company, partner, currency, date):
+                continue
+
+            declaration = self.env['l10n_it_edi_doi.declaration_of_intent']\
+                ._fetch_valid_declaration_of_intent(company, partner, currency, date)
+            order.l10n_it_edi_doi_id = declaration
+
+    @api.depends('tax_totals', 'invoice_ids', 'invoice_ids.state', 'invoice_ids.tax_totals')
+    def _compute_l10n_it_edi_doi_not_yet_invoiced(self):
+        for order in self:
+            declaration = order.l10n_it_edi_doi_id
+            order.l10n_it_edi_doi_not_yet_invoiced = order._l10n_it_edi_doi_get_amount_not_yet_invoiced(declaration)
+
+    @api.depends('l10n_it_edi_doi_id', 'l10n_it_edi_doi_id.remaining', 'state', 'tax_totals')
+    def _compute_l10n_it_edi_doi_warning(self):
+        for order in self:
+            order.l10n_it_edi_doi_warning = ''
+            declaration = order.l10n_it_edi_doi_id
+
+            show_warning = declaration and order.state != 'cancelled'
```

Comment:
As discussed above we need it for moves.

For SOs it is not really that problematic. But we should probably still need to cancel them in case the threshold is exceeded?